walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,95 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> output = Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+
+    // Update output time slots for all mandatory attendees
+    for (String attendee : request.getAttendees()) {
+      output = updateOutput(events, attendee, output);
+    }
+
+    List<TimeRange> mandatoryOutput = cleanUpOutput(request, output);
+    
+    // Update output time slots for all optional attendees
+    for (String attendee : request.getOptionalAttendees()) {
+      output = updateOutput(events, attendee, output);
+    }
+
+    List<TimeRange> optionalOutput = cleanUpOutput(request, output);
+
+    if (optionalOutput.size() == 0 && request.getAttendees().size() != 0) {
+      return mandatoryOutput;
+    }
+    return optionalOutput;
+  }
+
+  List<TimeRange> updateOutput(Collection<Event> events, String attendee, List<TimeRange> output) {
```

Comment:
Also maybe change this name to something more descriptive.  I'm thinking maybe `findAvailableTimeRanges` or something similar?

Reply:
Will do!